### ISSUE TITLE
Django's LiveServerTestCase is not supported

### ISSUE THREAD

**Issue by emperorcezar** [box/flaky#70]
Django uses a `__call__` method on the test for setup and teardown of the live server. This currently breaks with flaky. I have a pull request coming momentarily that addresses this.

**Comment by jmoldow** [comment]
FYI, the offending ``__call__`` definition is in Django's ``SimpleTestCase`` (which is a superclass of ``LiveServerTestCase``).

What exactly is the problem? Can you provide an example test, and the stack trace of the error that is produced?

The test's ``__call__`` is called in ``Test.runTest()`` (https://github.com/nose-devs/nose/blob/master/nose/case.py#L151), which is called in ``Test.run()`` (https://github.com/nose-devs/nose/blob/master/nose/case.py#L133), which is called in ``FlakyPlugin.stopTest()`` (https://github.com/box/flaky/blob/master/flaky/flaky_nose_plugin.py#L110). So I would expect that the Django setup and teardown would happen automatically. What actually happens?

**Comment by Jeff-Meadows** [comment]
Without more information about how this breaks, we won't be able to fix it or offer any advice. Please reopen if you have more information!

**Comment by amygdalama** [comment]
I'm running into this issue too. Hopefully this provides more clarity:

When hooking up the flaky nose plugin and running a single failing test that inherits from Django's TestCase, here's the order of events:
```
>>> _pre_setup
>>> test run #1
>>> flaky startTest
>>> flaky stopTest
>>> _pre_setup
>>> test run #2
>>> flaky startTest
>>> flaky stopTest
>>> _pre_setup
>>> test run #3
>>> flaky startTest
>>> flaky stopTest
>>> _post_teardown
>>> _post_teardown
>>> _post_teardown
```

The important piece here is that `_post_teardown` isn't called in between test retries. This causes a buffet of errors in our test suite. I can provide stacktraces if you'd like, but I'm not sure they'd be helpful.

One way to fix this could be to change how tests are re-run. We could get a handle on the suite that's running inside [`prepareTest`](http://nose.readthedocs.io/en/latest/plugins/interface.html#nose.plugins.base.IPluginInterface.prepareTest) and then when tests fail/error, conditionally add the test to the suite using [`addTest`](https://github.com/nose-devs/nose/blob/7c26ad1e6b7d308cafa328ad34736d34028c122a/nose/suite.py#L67-L68). I tested this out locally to make sure that the tests actually get re-run when added to the suite via `addTest`, and it's working so far. Would y'all consider this approach? If so, I'd happily try making the change.

**Comment by Jeff-Meadows** [comment]
Thanks @amygdalama for shedding some more light on the issue. To recap:

Django subclasses `unittest.TestCase` and breaks the plugin architecture by doing important things before `startTest` is called and after `stopTest` is called.

Flaky reruns tests with an error or failure when `stopTest` is called. This is supposed to be the point at which test teardown is complete, so it's safe to rerun the test.

The good news: nose does also define an `afterTest` hook that looks like it might be called even after Django's _post_teardown.

@amygdalama - would you mind testing this out? Change `def stopTest` to `def afterTest` in https://github.com/box/flaky/blob/master/flaky/flaky_nose_plugin.py#L98, maybe on your local machine. If this runs the tests the way you're hoping, I don't see an issue changing flaky to use that hook instead.

**Comment by amygdalama** [comment]
Yep, that works! It's definitely a smaller change than what I was suggesting, which I just pushed here: https://github.com/venmo/flaky/commit/6d1684132c04166f48ea5b79ed7b05f34677fa09

Let me run our test suite with the `afterTest` change to see if it plays nicely with the other plugins we have hooked up. I'm a little worried about that, since I haven't run our entire test suite with flaky yet. Fingers crossed.
